### PR TITLE
Checkout: Editor checkout modal.

### PR DESCRIPTION
**Pull request by yansern**
#### Changes proposed in this Pull Request

* Adds editor checkout modal.

This PR supercedes https://github.com/Automattic/wp-calypso/pull/45902 with the following differences:
- Calling the checkout modal is done through `window.wp.hooks.doAction`.
- `AsyncLoad` for checkout modal.
- Passing of the entire cart data as an object.
- The checkout modal is placed inside `client/blocks/editor-checkout-modal`.
- Hiding behind feature flag `gutenboarding/checkout-overlay`.

#### Testing instructions

* Sandbox widgets.wp.com.
* Run `yarn dev --sync` on `apps/wpcom-block-editor`.
* Open block editor of the free site with the feature flag `?flags=post-editor/checkout-overlay`, e.g. `calypso.localhost:3000/block-editor/post/your_sandboxed_free_site.wordpress.com?flags=post-editor/checkout-overlay`
* Open up dev console, change the frame to `post.php`, and run the following code.
```js
let cartData = { products: [{
    product_id: 1009,
    product_slug: 'personal-bundle',
}] };

window.wp.hooks.doAction('a8c.wpcom-block-editor.openCheckoutModal', cartData);
```
* Checkout modal should appear.

### CODE REVIEW

File: client/blocks/editor-checkout-modal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+const wpcom = wp.undocumented();
+
+class EditorCheckoutModal extends Component {
+	static propTypes = {
+		site: PropTypes.object,
+		isOpen: PropTypes.bool,
+		onClose: PropTypes.func,
+		cartData: PropTypes.object,
+	};
```
Comment (alshakero) [inline client/blocks/editor-checkout-modal/index.tsx:31]: I think PropTypes aren't needed in TypeScript files :D

File: client/gutenberg/editor/calypsoify-iframe.tsx [hunk 2]
```diff
@@ -708,6 +725,14 @@ class CalypsoifyIframe extends Component<
...
 					visible={ isMediaModalVisible }
 				/>
+				{ hasCheckoutOverlay && (
+					<AsyncLoad
+						require="blocks/editor-checkout-modal"
+						onClose={ this.closeCheckoutModal }
+						isOpen={ isCheckoutModalVisible }
+						cartData={ cartData }
+					/>
+				) }
```
Comment (alshakero) [inline client/gutenberg/editor/calypsoify-iframe.tsx:735]: I was thinking about this yesterday. This is loaded async indeed. But do we need to load it for everyone who may or may not click "Upgrade"?

I think this is worth a try and see how it feels: 

```jsx
{ hasCheckoutOverlay && isCheckoutModalVisible (
	<AsyncLoad
		require="blocks/editor-checkout-modal"
		onClose={ this.closeCheckoutModal }
		isOpen
		cartData={ cartData }
	/>
) }
```

**Review by alshakero — APPROVED** [review]
So freaking neat. I love it!

File: client/gutenberg/editor/calypsoify-iframe.tsx (hunk 2 above)
Comment (yansern) [reply client/gutenberg/editor/calypsoify-iframe.tsx:735]: Yup true! This ensure it is only loaded when it really needs to be visible at the trade off of a bit of load time. Let's see how the load time feels like.

File: client/blocks/editor-checkout-modal/index.tsx (hunk 1 above)
Comment (yansern) [reply client/blocks/editor-checkout-modal/index.tsx:31]: Perils of copy & paste!

File: client/blocks/editor-checkout-modal/index.tsx [hunk 3]
```diff
@@ -22,23 +21,36 @@ import './style.scss';
...
-		isOpen: PropTypes.bool,
-		onClose: PropTypes.func,
-		cartData: PropTypes.object,
-	};
+interface CartData {
+	products: Array< {
+		product_id: number;
+		product_slug: string;
+	} >;
+}
```
Comment (yansern) [inline client/blocks/editor-checkout-modal/index.tsx:29]: Okay so I wonder if there is already an interface written for `CartData` somewhere, not a blocker for this PR, something to look up later.

File: client/gutenberg/editor/calypsoify-iframe.tsx (hunk 2 above)
Comment (simison) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: Same convo here: https://github.com/Automattic/wp-calypso/pull/45932#discussion_r495781898 — Just FYI.

File: client/blocks/editor-checkout-modal/style.scss [hunk 4]
```diff
@@ -0,0 +1,24 @@
+.editor-checkout-modal {
+	position: fixed;
+	top: 0;
+	right: -100%;
+	z-index: 999999;
```
Comment (simison) [inline client/blocks/editor-checkout-modal/style.scss:5]: Nit; there is a SASS mixin you can use for this kind of thing. :-)

Comment (yansern) [reply client/blocks/editor-checkout-modal/style.scss:5]: Oh really, I've been... blessed... with working with modals and popovers and other things that float around the screen, that SASS mixin might just save me. ;)

**Pull request merged**